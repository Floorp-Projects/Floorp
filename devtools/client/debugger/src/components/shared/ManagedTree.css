/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

.managed-tree .tree {
  user-select: none;

  white-space: nowrap;
  overflow: auto;
  min-width: 100%;

  display: grid;
  grid-template-columns: 1fr;
  align-content: start;

  line-height: 1.4em;
}

.managed-tree .tree button {
  display: block;
}

.managed-tree .tree .node {
  padding: 2px 3px 2px 3px;
  position: relative;
}

.managed-tree .tree .node.focused {
  color: var(--theme-selection-color);
  background-color: var(--theme-selection-background);
}

html:not([dir="rtl"]) .managed-tree .tree .node > div {
  margin-left: 10px;
}

html[dir="rtl"] .managed-tree .tree .node > div {
  margin-right: 10px;
}

.managed-tree .tree-node button {
  position: fixed;
}
