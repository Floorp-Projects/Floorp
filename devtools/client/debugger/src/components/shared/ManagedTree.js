/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { Component } from "react";
import "./ManagedTree.css";

const Tree = require("devtools/client/shared/components/Tree");

class ManagedTree extends Component {
  constructor(props) {
    super(props);
    this.state = {
      expanded: props.expanded || new Set(),
    };
  }

  static defaultProps = {
    onFocus: () => {},
  };

  componentWillReceiveProps(nextProps) {
    const { listItems, highlightItems } = this.props;
    if (nextProps.listItems && nextProps.listItems != listItems) {
      this.expandListItems(nextProps.listItems);
    }

    if (
      nextProps.highlightItems &&
      nextProps.highlightItems != highlightItems &&
      nextProps.highlightItems.length
    ) {
      this.highlightItem(nextProps.highlightItems);
    }
  }

  setExpanded = (item, isExpanded, shouldIncludeChildren) => {
    const expandItem = i => {
      const path = this.props.getPath(i);
      if (isExpanded) {
        expanded.add(path);
      } else {
        expanded.delete(path);
      }
    };
    const { expanded } = this.state;
    expandItem(item);

    if (shouldIncludeChildren) {
      let parents = [item];
      while (parents.length) {
        const children = [];
        for (const parent of parents) {
          if (parent.contents?.length) {
            for (const child of parent.contents) {
              expandItem(child);
              children.push(child);
            }
          }
        }
        parents = children;
      }
    }
    this.setState({ expanded });

    if (isExpanded && this.props.onExpand) {
      this.props.onExpand(item, expanded);
    } else if (!isExpanded && this.props.onCollapse) {
      this.props.onCollapse(item, expanded);
    }
  };

  expandListItems(listItems) {
    const { expanded } = this.state;
    listItems.forEach(item => expanded.add(this.props.getPath(item)));
    this.props.onFocus(listItems[0]);
    this.setState({ expanded });
  }

  highlightItem(highlightItems) {
    const { expanded } = this.state;
    // This file is visible, so we highlight it.
    if (expanded.has(this.props.getPath(highlightItems[0]))) {
      this.props.onFocus(highlightItems[0]);
    } else {
      // Look at folders starting from the top-level until finds a
      // closed folder and highlights this folder
      const index = highlightItems
        .reverse()
        .findIndex(
          item =>
            !expanded.has(this.props.getPath(item)) && item.name !== "root"
        );

      if (highlightItems[index]) {
        this.props.onFocus(highlightItems[index]);
      }
    }
  }

  render() {
    const { expanded } = this.state;
    return (
      <div className="managed-tree">
        <Tree
          {...this.props}
          isExpanded={item => expanded.has(this.props.getPath(item))}
          focused={this.props.focused}
          getKey={this.props.getPath}
          onExpand={(item, shouldIncludeChildren) =>
            this.setExpanded(item, true, shouldIncludeChildren)
          }
          onCollapse={(item, shouldIncludeChildren) =>
            this.setExpanded(item, false, shouldIncludeChildren)
          }
          onFocus={this.props.onFocus}
          renderItem={(...args) =>
            this.props.renderItem(...args, {
              setExpanded: this.setExpanded,
            })
          }
        />
      </div>
    );
  }
}

export default ManagedTree;
