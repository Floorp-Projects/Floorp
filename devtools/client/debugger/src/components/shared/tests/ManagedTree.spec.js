/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { mount, shallow } from "enzyme";

import ManagedTree from "../ManagedTree";

function getTestContent() {
  const testTree = {
    a: {
      value: "FOO",
      children: [
        { value: 1 },
        { value: 2 },
        { value: 3 },
        { value: 4 },
        { value: 5 },
      ],
    },
    b: {
      value: "BAR",
      children: [
        { value: "A" },
        { value: "B" },
        { value: "C" },
        { value: "D" },
        { value: "E" },
      ],
    },
    c: { value: "BAZ" },
  };
  const renderItem = item => <div>{item.value ? item.value : item}</div>;
  const onFocus = jest.fn();
  const onExpand = jest.fn();
  const onCollapse = jest.fn();
  const getPath = (item, i) => {
    if (item.value) {
      return item.value;
    }
    if (i) {
      return `${i}`;
    }
    return `${item}-$`;
  };

  return {
    testTree,
    props: {
      getRoots: () => Object.keys(testTree),
      getParent: item => null,
      getChildren: branch => branch.children || [],
      itemHeight: 24,
      autoExpandAll: true,
      autoExpandDepth: 1,
      getPath,
      renderItem,
      onFocus,
      onExpand,
      onCollapse,
    },
  };
}

describe("ManagedTree", () => {
  it("render", () =>
    expect(
      shallow(<ManagedTree {...getTestContent().props} />)
    ).toMatchSnapshot());
  it("expands list items", () => {
    const { props, testTree } = getTestContent();
    const wrapper = shallow(<ManagedTree {...props} />);
    wrapper.setProps({
      listItems: testTree.b.children,
    });
    expect(wrapper).toMatchSnapshot();
  });
  it("highlights list items", () => {
    const { props, testTree } = getTestContent();
    const wrapper = shallow(<ManagedTree {...props} />);
    wrapper.setProps({
      highlightItems: testTree.a.children,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("sets expanded items", () => {
    const { props, testTree } = getTestContent();
    const wrapper = mount(<ManagedTree {...props} />);
    expect(wrapper).toMatchSnapshot();
    // We auto-expanded the first layer, so unexpand first node.
    wrapper
      .find("TreeNode")
      .first()
      .simulate("click");
    expect(wrapper).toMatchSnapshot();
    expect(props.onExpand).toHaveBeenCalledWith(
      "c",
      new Set(
        Object.keys(testTree)
          .filter(i => i !== "a")
          .map(k => `${k}-$`)
      )
    );
    wrapper
      .find("TreeNode")
      .first()
      .simulate("click");
    expect(props.onExpand).toHaveBeenCalledWith(
      "c",
      new Set(Object.keys(testTree).map(k => `${k}-$`))
    );
  });
});
