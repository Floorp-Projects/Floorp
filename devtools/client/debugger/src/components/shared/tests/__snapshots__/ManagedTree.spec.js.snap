// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManagedTree expands list items 1`] = `
<div
  className="managed-tree"
>
  <Tree
    autoExpandAll={true}
    autoExpandDepth={1}
    getChildren={[Function]}
    getKey={[Function]}
    getParent={[Function]}
    getPath={[Function]}
    getRoots={[Function]}
    isExpanded={[Function]}
    itemHeight={24}
    listItems={
      Array [
        Object {
          "value": "A",
        },
        Object {
          "value": "B",
        },
        Object {
          "value": "C",
        },
        Object {
          "value": "D",
        },
        Object {
          "value": "E",
        },
      ]
    }
    onCollapse={[Function]}
    onExpand={[Function]}
    onFocus={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "value": "A",
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    renderItem={[Function]}
  />
</div>
`;

exports[`ManagedTree highlights list items 1`] = `
<div
  className="managed-tree"
>
  <Tree
    autoExpandAll={true}
    autoExpandDepth={1}
    getChildren={[Function]}
    getKey={[Function]}
    getParent={[Function]}
    getPath={[Function]}
    getRoots={[Function]}
    highlightItems={
      Array [
        Object {
          "value": 5,
        },
        Object {
          "value": 4,
        },
        Object {
          "value": 3,
        },
        Object {
          "value": 2,
        },
        Object {
          "value": 1,
        },
      ]
    }
    isExpanded={[Function]}
    itemHeight={24}
    onCollapse={[Function]}
    onExpand={[Function]}
    onFocus={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "value": 5,
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    renderItem={[Function]}
  />
</div>
`;

exports[`ManagedTree render 1`] = `
<div
  className="managed-tree"
>
  <Tree
    autoExpandAll={true}
    autoExpandDepth={1}
    getChildren={[Function]}
    getKey={[Function]}
    getParent={[Function]}
    getPath={[Function]}
    getRoots={[Function]}
    isExpanded={[Function]}
    itemHeight={24}
    onCollapse={[Function]}
    onExpand={[Function]}
    onFocus={[MockFunction]}
    renderItem={[Function]}
  />
</div>
`;

exports[`ManagedTree sets expanded items 1`] = `
<ManagedTree
  autoExpandAll={true}
  autoExpandDepth={1}
  getChildren={[Function]}
  getParent={[Function]}
  getPath={[Function]}
  getRoots={[Function]}
  itemHeight={24}
  onCollapse={[MockFunction]}
  onExpand={
    [MockFunction] {
      "calls": Array [
        Array [
          "a",
          Set {
            "a-$",
            "b-$",
            "c-$",
          },
        ],
        Array [
          "b",
          Set {
            "a-$",
            "b-$",
            "c-$",
          },
        ],
        Array [
          "c",
          Set {
            "a-$",
            "b-$",
            "c-$",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onFocus={[MockFunction]}
  renderItem={[Function]}
>
  <div
    className="managed-tree"
  >
    <Tree
      autoExpandAll={true}
      autoExpandDepth={1}
      getChildren={[Function]}
      getKey={[Function]}
      getParent={[Function]}
      getPath={[Function]}
      getRoots={[Function]}
      isExpanded={[Function]}
      itemHeight={24}
      onCollapse={[Function]}
      onExpand={[Function]}
      onFocus={[MockFunction]}
      renderItem={[Function]}
    >
      <div
        className="tree "
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        onKeyUp={[Function]}
        role="tree"
        style={Object {}}
        tabIndex="0"
      >
        <TreeNode
          active={false}
          depth={0}
          expanded={true}
          focused={false}
          id="a-$"
          index={0}
          isExpandable={false}
          item="a"
          key="a-$-inactive"
          onClick={[Function]}
          onCollapse={[Function]}
          onExpand={[Function]}
          renderItem={[Function]}
        >
          <div
            aria-expanded={true}
            aria-level={1}
            className="tree-node"
            data-expandable={false}
            id="a-$"
            onClick={[Function]}
            onKeyDownCapture={null}
            role="treeitem"
          >
            <div>
              a
            </div>
          </div>
        </TreeNode>
        <TreeNode
          active={false}
          depth={0}
          expanded={true}
          focused={false}
          id="1"
          index={1}
          isExpandable={false}
          item="b"
          key="1-inactive"
          onClick={[Function]}
          onCollapse={[Function]}
          onExpand={[Function]}
          renderItem={[Function]}
        >
          <div
            aria-expanded={true}
            aria-level={1}
            className="tree-node"
            data-expandable={false}
            id="1"
            onClick={[Function]}
            onKeyDownCapture={null}
            role="treeitem"
          >
            <div>
              b
            </div>
          </div>
        </TreeNode>
        <TreeNode
          active={false}
          depth={0}
          expanded={true}
          focused={false}
          id="2"
          index={2}
          isExpandable={false}
          item="c"
          key="2-inactive"
          onClick={[Function]}
          onCollapse={[Function]}
          onExpand={[Function]}
          renderItem={[Function]}
        >
          <div
            aria-expanded={true}
            aria-level={1}
            className="tree-node"
            data-expandable={false}
            id="2"
            onClick={[Function]}
            onKeyDownCapture={null}
            role="treeitem"
          >
            <div>
              c
            </div>
          </div>
        </TreeNode>
      </div>
    </Tree>
  </div>
</ManagedTree>
`;

exports[`ManagedTree sets expanded items 2`] = `
<ManagedTree
  autoExpandAll={true}
  autoExpandDepth={1}
  getChildren={[Function]}
  getParent={[Function]}
  getPath={[Function]}
  getRoots={[Function]}
  itemHeight={24}
  onCollapse={
    [MockFunction] {
      "calls": Array [
        Array [
          "a",
          Set {
            "b-$",
            "c-$",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onExpand={
    [MockFunction] {
      "calls": Array [
        Array [
          "a",
          Set {
            "b-$",
            "c-$",
          },
        ],
        Array [
          "b",
          Set {
            "b-$",
            "c-$",
          },
        ],
        Array [
          "c",
          Set {
            "b-$",
            "c-$",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onFocus={
    [MockFunction] {
      "calls": Array [
        Array [
          "a",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  renderItem={[Function]}
>
  <div
    className="managed-tree"
  >
    <Tree
      autoExpandAll={true}
      autoExpandDepth={1}
      getChildren={[Function]}
      getKey={[Function]}
      getParent={[Function]}
      getPath={[Function]}
      getRoots={[Function]}
      isExpanded={[Function]}
      itemHeight={24}
      onCollapse={[Function]}
      onExpand={[Function]}
      onFocus={
        [MockFunction] {
          "calls": Array [
            Array [
              "a",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      renderItem={[Function]}
    >
      <div
        className="tree "
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        onKeyUp={[Function]}
        role="tree"
        style={Object {}}
        tabIndex="0"
      >
        <TreeNode
          active={false}
          depth={0}
          expanded={false}
          focused={false}
          id="a-$"
          index={0}
          isExpandable={false}
          item="a"
          key="a-$-inactive"
          onClick={[Function]}
          onCollapse={[Function]}
          onExpand={[Function]}
          renderItem={[Function]}
        >
          <div
            aria-level={1}
            className="tree-node"
            data-expandable={false}
            id="a-$"
            onClick={[Function]}
            onKeyDownCapture={null}
            role="treeitem"
          >
            <div>
              a
            </div>
          </div>
        </TreeNode>
        <TreeNode
          active={false}
          depth={0}
          expanded={true}
          focused={false}
          id="1"
          index={1}
          isExpandable={false}
          item="b"
          key="1-inactive"
          onClick={[Function]}
          onCollapse={[Function]}
          onExpand={[Function]}
          renderItem={[Function]}
        >
          <div
            aria-expanded={true}
            aria-level={1}
            className="tree-node"
            data-expandable={false}
            id="1"
            onClick={[Function]}
            onKeyDownCapture={null}
            role="treeitem"
          >
            <div>
              b
            </div>
          </div>
        </TreeNode>
        <TreeNode
          active={false}
          depth={0}
          expanded={true}
          focused={false}
          id="2"
          index={2}
          isExpandable={false}
          item="c"
          key="2-inactive"
          onClick={[Function]}
          onCollapse={[Function]}
          onExpand={[Function]}
          renderItem={[Function]}
        >
          <div
            aria-expanded={true}
            aria-level={1}
            className="tree-node"
            data-expandable={false}
            id="2"
            onClick={[Function]}
            onKeyDownCapture={null}
            role="treeitem"
          >
            <div>
              c
            </div>
          </div>
        </TreeNode>
      </div>
    </Tree>
  </div>
</ManagedTree>
`;
